### PR TITLE
Feature scraper

### LINKED ISSUE #4: Basic scraping walker
Basic addition of all content by traversing:

- homepage
- Special:CategoryListing or Special:Sitemap
- categories listed inside main categories (should walk through category graph recursively)
- Articles
- Support content:
    - About (footer)
    - Experts (footer)
    - Delivering-a-Trustworthy-Experience
    - Mission (About)
    - Hybrid-Organization (About)
    - Powered-and-Inspired-by-MediaWiki (About)
    - Tour
- all links rewritten
    - internal: to in-zim path
    - external: external icon ; removed if requested

### LINKED ISSUE #5: Article Template
- Pre-content: Categories breadcrumb, title
- NO pdf button
- Intro: Table Of Content, Author, Description
- Content without PDF buttons
- Q&A with scores but not actionnable nor «Ask a question block»
- Tips if any
- Video if any
- Relateds
- About this article:
    - score not actionnable
    - without interwiki links (langs)
    - wihout print/mail
- wihout «is this article up to date?»

### PR DESCRIPTION
**Pull request by ElenaKusevska**

### CODE REVIEW

File: wikihow2zim/scraper.py [hunk 1]
```diff
@@ -76,12 +76,161 @@ def add_assets(self):
...
+        # Get all the articles by walking through all the main categories
+        # and the sub-categories
+
+        self.articles_list = {}  # keep track of each article processed and
+        # where each article was referenced from
+        self.subcat_list = {}  # keep track of each subcategory processed, and
+        # where each subcategory was referenced from
+
+        main_wiki_url = "https://www.wikihow.com"
+        sitemap_url = "https://www.wikihow.com/Special:Sitemap"
```
Comment (rgaudin) [inline wikihow2zim/scraper.py]: No need to hard-code wikiHow EN. I believe this page is available at same path on all websites

File: wikihow2zim/scraper.py [hunk 2]
```diff
@@ -76,12 +76,161 @@ def add_assets(self):
...
+                sups = text.find_all("sup")
+                for ss in sups:
+                    ss.decompose()
+                step["text"] = text
+
+                method["steps"].append(step)
+
+            methods.append(method)
+
+        sys.exit()
```
Comment (rgaudin) [inline wikihow2zim/scraper.py]: `sys` is not imported. Again, you should integrate `flake8` in your text editor or run it periodically as you have several undeclared variables in this code

File: wikihow2zim/scraper.py [hunk 3]
```diff
@@ -76,12 +76,161 @@ def add_assets(self):
...
+            logger.critical(f"Unable to retrieve homepage at {url}: {exc}")
+            logger.exception(exc)
+
+        soup = bs4.BeautifulSoup(response.text, "html.parser")
+
+        # Find main categories
+        category_divs = soup.find_all("div", {"class": "cat_list"})
+        self.main_categories = []
+        for div in category_divs:
+            h3 = div.find("h3")
```
Comment (rgaudin) [inline wikihow2zim/scraper.py]: You may want to use a different selector to directly retrieve the `h3`

File: wikihow2zim/scraper.py [hunk 4]
```diff
@@ -76,12 +76,161 @@ def add_assets(self):
...
+            cat_url = main_wiki_url + h3.find("a")["href"]
+            print("0 category " + cat_url, flush=True)
+            current_category = [h3.text, self.output_dir + "/C" + h3.find("a")["href"]]
```
Comment (rgaudin) [inline wikihow2zim/scraper.py]: This looks like you are trying to replicate the kiwix-serve path here.

When you add path `home` to a zim (`.add_content_for(path="home", content="hello")`), it is reachable by kiwix-serve at `http://{host}:{port}/{zim_name}/C/home` or even without the `C/` if you are using the latest (unreleased I believe) kiwix-serve.
You **should not** care about this prefix as this is a kiwix-serve only prefix that is used to seggregate zim files (kiwix-serve can serve multiple zim files).

File: wikihow2zim/scraper.py [hunk 5]
```diff
@@ -76,12 +76,161 @@ def add_assets(self):
...
+        # )
+
+    def walk_subcategories(self, main_wiki_url, cat_url, recursion_depth):
+        try:
+            response = requests.get(cat_url)
+        except Exception as exc:
+            logger.critical(f"Unable to retrieve homepage at {cat_url}: {exc}")
+            logger.exception(exc)
+
+        soup = bs4.BeautifulSoup(response.text, "html.parser")
```
Comment (rgaudin) [inline wikihow2zim/scraper.py]: Already mentioned that we want to use `lxml` parser everywhere. `html.parser` is too slow.

File: wikihow2zim/scraper.py [hunk 6]
```diff
@@ -76,12 +76,161 @@ def add_assets(self):
...
+        article_divs = soup.find_all("div", {"class": "responsive_thumb"})
+        for article_div in article_divs:
+            article_url = article_div.find("a")["href"]
+            if not article_url in self.articles_list:
```
Comment (rgaudin) [inline wikihow2zim/scraper.py]: convention is to use `if xxx not in yyy` not `if not xxx in yyy`.

Also, the is a poorly chosen name for a `dict`: `_list` implies a type which is different.

File: wikihow2zim/scraper.py [hunk 7]
```diff
@@ -76,12 +76,161 @@ def add_assets(self):
...
+        # )
+        # self.creator.add_item(
+        #     URLItem(
+        #         url=url,
+        #         path="assets/static/wikihow_logo.png",
+        #         mimetype="image/png",
+        #     )
+        # )
+
+    def walk_subcategories(self, main_wiki_url, cat_url, recursion_depth):
```
Comment (rgaudin) [inline wikihow2zim/scraper.py]: Why would we need to pass the static `main_wiki_url` ?

File: wikihow2zim/scraper.py [hunk 8]
```diff
@@ -76,12 +76,161 @@ def add_assets(self):
...
+        except Exception as exc:
+            logger.critical(f"Unable to retrieve homepage at {cat_url}: {exc}")
+            logger.exception(exc)
+
+        soup = bs4.BeautifulSoup(response.text, "html.parser")
+        article_divs = soup.find_all("div", {"class": "responsive_thumb"})
+        for article_div in article_divs:
+            article_url = article_div.find("a")["href"]
+            if not article_url in self.articles_list:
+                self.articles_list[article_url] = [str(recursion_depth) + " " + cat_url]
```
Comment (rgaudin) [inline wikihow2zim/scraper.py]: what's the purpose of this ? Debug print only?

File: wikihow2zim/scraper.py [hunk 9]
```diff
@@ -76,12 +76,161 @@ def add_assets(self):
...
+                    str(recursion_depth) + " " + cat_url
+                )
+            print(str(recursion_depth) + " category " + subcat_url, flush=True)
+            self.walk_subcategories(main_wiki_url, subcat_url, recursion_depth + 1)
+
+    def walk_categories(self):
+        # Get all the articles by walking through all the main categories
+        # and the sub-categories
+
+        self.articles_list = {}  # keep track of each article processed and
```
Comment (rgaudin) [inline wikihow2zim/scraper.py]: Don't create scraper instance properties inside methods if those are to be used anyway. This should be moved to __init__

**Review by rgaudin — COMMENTED** [review]
A few hints on your WIP

File: wikihow2zim/scraper.py (hunk 1 above)
Comment (ElenaKusevska) [reply wikihow2zim/scraper.py]: @rgaudin, you're right, I just realized it's just self.url

File: wikihow2zim/scraper.py (hunk 2 above)
Comment (ElenaKusevska) [reply wikihow2zim/scraper.py]: @rgaudin, I don't intend on using sys. This is just a draft. I'm just using this line to mark where I'm working on. I'll delete it.

File: wikihow2zim/scraper.py (hunk 8 above)
Comment (ElenaKusevska) [reply wikihow2zim/scraper.py]: Yes, I used this to see how deep the recursion goes, and also how many places an article is accessed from. I can delete these lines when I'm done.

File: wikihow2zim/scraper.py (hunk 9 above)
Comment (ElenaKusevska) [reply wikihow2zim/scraper.py]: @rgaudin Let's Keep them for now. We can remove them later on.

File: wikihow2zim/scraper.py (hunk 4 above)
Comment (ElenaKusevska) [reply wikihow2zim/scraper.py]: @rgaudin This isn't for the path when creating the zim, it's for the link. I realized that in-zim links don't work unless the entire path is written with kiwix-serve

Comment (ElenaKusevska) [reply wikihow2zim/scraper.py]: Nevermind, I figured it out :)

**Pull request merged**